oximeter/collector/src/agent.rs

```diff
@@ -166,35 +295,112 @@ async fn collection_task(
+                                            queue is closed. Attempting to \
+                                            notify caller and exiting.",
+                                        );
+                                        let _ = tok.send(Ok(()));
+                                        return;
+                                    }
+                                    TrySendError::Full(tok) => {
+                                        error!(
+                                            log,
+                                            "collection task forced collection \
+                                            queue is full! This should never \
+                                            happen, and probably indicates \
+                                            a bug in your test code, such as \
+                                            calling `force_collection()` many \
+                                            times"
+                                        );
+                                        if tok
+                                            .send(Err(ForcedCollectionQueueFull))
+                                            .is_err()
+                                        {
+                                            warn!(
+                                                log,
+                                                "failed to notify caller of \
+                                                force_collection(), oneshot is \
+                                                closed"
+                                            );
+                                        }
+                                    }
+                                }
+                            }
+                        }
                     },
                     Some(CollectionMessage::Update(new_info)) => {
-                        producer = new_info;
                         debug!(
                             log,
                             "collection task received request to update its producer information";
-                            "interval" => ?producer.interval,
-                            "address" => producer.address,
+                            "interval" => ?new_info.interval,
+                            "address" => new_info.address,
                         );
 
-                        // Update the logger with the new information as well.
-                        log = orig_log.new(o!("address" => producer.address));
-                        collection_timer = interval(producer.interval);
+                        // Update our own collection timer, and notify the
+                        // collection task of the new data as well.
+                        collection_timer = interval(new_info.interval);
                         collection_timer.tick().await; // completes immediately
+                        if let Err(_) = producer_info_tx.send(new_info) {
```

Comment:
To combine this a bit with #7126: we could maybe use [send_if_modified](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html#method.send_if_modified) instead? It can't fail and also gives a natural place to check whether the new info is actually a change (and therefore skip the timer rebuild).

Reply:
That sounds good. I'm not familiar with that interface, so I'll need to read up a bit. I was going to merge #7126 first anyway, and so I can make sure to handle conditional modification here too. Thanks for the tip!

Reply:
Reworked to use `send_if_modified()` in 010e3a273.